lib/src/services/song_recognition.dart

```diff
@@ -0,0 +1,193 @@
+      final sink = outputFile.openWrite();
+
+      streamSubscription = response.stream.listen((chunk) async {
+        final bytes = await outputFile.length();
+
+        if (bytes > 0) {
+          // we can calculate the duration by using the bitrate and
+          //the file size, the formula is found here:
+          // http://www.audiomountain.com/tech/audio-file-size.html
+          final bytePerSecond = bitRate / 8 * 1000;
+          final expectedBytes = bytePerSecond * durationInSeconds;
+
+          if (expectedBytes < bytes) {
+            await sink.flush();
+            await sink.close();
+
+            await streamSubscription?.cancel();
+            if (!completer.isCompleted) {
+              completer.complete(outputFile);
+            }
+            return;
+          }
+        }
+
+        sink.add(chunk);
+      });
+    } catch (_) {
+      completer.completeError(
+        Exception(
+          'There was a problem generating the file',
+        ),
+      );
+    }
+    return completer.future;
+  }
+
+  /// Identifies the current song playing in the radio.
+  ///
+  /// Receives a [radioId] to identify the radio.
+  Future<String> identify(String radioId) async {
+    final stopwatch = Stopwatch()..start();
+
+    final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
+
+    final stringToSign = buildStringToSign(
+      'POST',
+      _options.endpoint,
+      _options.accessKey,
+      _options.dataType,
+      _options.signatureVersion,
+      timestamp.toString(),
+    );
+    final signature = sign(stringToSign, _options.accessSecret);
+
+    final uri = Uri(
+      scheme: 'https',
+      host: arcCloudHost,
+      path: 'v1/identify',
+    );
+
+    try {
+      final songFile = await generateSample(radioId: radioId);
+
+      final sample = songFile.readAsBytesSync()..buffer.asUint8List();
+
+      final fields = {
+        'access_key': _options.accessKey,
+        'data_type': _options.dataType,
+        'signature_version': _options.signatureVersion,
+        'signature': signature,
+        'sample_bytes': sample.length.toString(),
+        'timestamp': timestamp.toString(),
+      };
+
+      final request = http.MultipartRequest('POST', uri);
+      request.fields.addAll(fields);
+      request.files.add(http.MultipartFile.fromBytes('sample', sample));
+
+      final streamedResponse = await request.send();
+      final response = await http.Response.fromStream(streamedResponse);
+      final result = SongRecognitionResponse.fromJson(
+        jsonDecode(response.body) as Map,
+      );
+
+      log('Done in ${stopwatch.elapsedMilliseconds}ms');
+      final track = result.metadata?.music?.first;
+
+      // Deletes the file after the recognition is done
+      songFile.deleteSync();
+
+      return track == null
+          ? "Couldn't recognise the song :("
+          : 'Song found: ${'${track.title} - ${track.artists?.first.name}'}';
```

Comment:
I think we could return the ´result´ itself. Also, the file deletion should be done externally, as including it in this function would be a mixing responsibility.